march_joint_inertia_controller/src/joint_inertia_controller.cpp

```diff
@@ -1,34 +1,175 @@
+{
+  pid_controller_.getGains(p, i, d, i_max, i_min, antiwindup);
+}
+
+void InertiaController::getGains(double& p, double& i, double& d, double& i_max, double& i_min)
+{
+  bool dummy;
+  pid_controller_.getGains(p, i, d, i_max, i_min, dummy);
+}
+
+void InertiaController::printDebug()
+{
+  pid_controller_.printValues();
+}
+
+std::string InertiaController::getJointName()
+{
+  return joint_.getName();
+}
+
+double InertiaController::getPosition()
+{
+  return joint_.getPosition();
+}
+
 void InertiaController::starting(const ros::Time& /* time */)
 {
-  init_pos_ = joint_.getPosition();
+  double pos_command = joint_.getPosition();
+
+  command_struct_.position = pos_command;
+  command_struct_.has_velocity = false;
+
+  command_.initRT(command_struct_);
+
+  pid_controller_.reset();
 }
 
-void InertiaController::update(const ros::Time& /* time */, const ros::Duration& /* period */)
+void InertiaController::update(const ros::Time& /* time */, const ros::Duration& period)
 {
-  double dpos = init_pos_ + 10 * sin(ros::Time::now().toSec());
-  double cpos = joint_.getPosition();
-  joint_.setCommand(-10 * (cpos - dpos));
+  command_struct_ = *(command_.readFromRT());
+  double command_position = command_struct_.position;
+  double command_velocity = command_struct_.velocity;
+  bool has_velocity_ = command_struct_.has_velocity;
+
+  double error, vel_error;
+  double commanded_effort;
+
+  double current_position = joint_.getPosition();
+
+  if (joint_urdf_->type == urdf::Joint::REVOLUTE)
+  {
+    angles::shortest_angular_distance_with_large_limits(current_position, command_position, joint_urdf_->limits->lower,
+                                                        joint_urdf_->limits->upper, error);
+  }
+  else if (joint_urdf_->type == urdf::Joint::CONTINUOUS)
+  {
+    error = angles::shortest_angular_distance(current_position, command_position);
+  }
+  else  // prismatic
+  {
+    error = command_position - current_position;
+  }
+
+  // Decide which of the two PID computeCommand() methods to call
+  if (has_velocity_)
+  {
+    // Compute velocity error if a non-zero velocity command was given
+    vel_error = command_velocity - joint_.getVelocity();
+
+    // Set the PID error and compute the PID command with nonuniform
+    // time step size. This also allows the user to pass in a precomputed derivative error.
+    commanded_effort = pid_controller_.computeCommand(error, vel_error, period);
+  }
+  else
+  {
+    // Set the PID error and compute the PID command with nonuniform
+    // time step size.
+    commanded_effort = pid_controller_.computeCommand(error, period);
+  }
+
+  inertia_estimator_.fill_buffers(joint_.getVelocity(), joint_.getEffort(), period);
+  inertia_estimator_.inertia_estimate();
+  inertia_estimator_.publishInertia();
+  // TO DO: Provide lookup table for gain selection
+  // TO DO: apply PID control
```

Comment:
I did not find JIRA issues for this. I think you should add them, because otherwise they will be forgotten.

Reply:
see my other comment